pkg/cgrouprate/cgrouprate.go

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Tetragon
+
+package cgrouprate
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"github.com/cilium/tetragon/pkg/logger"
+	"github.com/cilium/tetragon/pkg/reader/notify"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	handle     *CgroupRate
+	handleLock sync.RWMutex
+)
+
+type Listener interface {
+	Notify(msg notify.Message) error
+	io.Closer
+}
+
+type CgroupRate struct {
+	listeners map[Listener]struct{}
+	log       logrus.FieldLogger
+}
+
+func newCgroupRate() *CgroupRate {
+	return &CgroupRate{
+		listeners: make(map[Listener]struct{}),
+		log:       logger.GetLogger(),
+	}
+}
+
+func NewCgroupRate(ctx context.Context) *CgroupRate {
+	handleLock.Lock()
+	defer handleLock.Unlock()
+
+	handle = newCgroupRate()
+	go handle.process(ctx)
+	return handle
+}
+
+func (r *CgroupRate) AddListener(listener Listener) {
+	r.listeners[listener] = struct{}{}
+}
```

Comment:
We seem to be using the same pattern as the observer, which is:

```
	rate := cgrouprate.NewCgroupRate(ctx, base.CgroupRateMap, &option.Config.CgroupRate)
	if rate != nil {
		cgrouprate.Config(base.CgroupRateOptionsMap)
		rate.AddListener(pm)
	}
```

This seems too complicated to me. Is there a reason for why we cannot have a single listener:

```
type CgroupRate struct {
	listener observer.Listener
	log      logrus.FieldLogger
	ch       chan *cgroupQueue
	opts     *option.CgroupRate
	hash     *program.Map
	cgroups  map[uint64]string
}
```

And do a single call where we initialize everything? I don't see any other use of AddListener but maybe I missed it?

Reply:
right, there's just one listener, so that can be simplified, will do that, thanks